pystac/utils.py

```diff
@@ -61,27 +138,78 @@ def make_relative_href(
         parent, then source_href will be returned unchanged.
     """
 
-    parsed_source = _urlparse(source_href)
-    parsed_start = _urlparse(start_href)
+    parsed_source = safe_urlparse(source_href)
+    parsed_start = safe_urlparse(start_href)
     if not (
         parsed_source.scheme == parsed_start.scheme
         and parsed_source.netloc == parsed_start.netloc
     ):
         return source_href
 
-    is_path = parsed_start.scheme == ""
-
-    if start_is_dir:
-        start_dir = parsed_start.path
+    if JoinType.from_parsed_uri(parsed_start) == JoinType.PATH:
+        return _make_relative_href_path(parsed_source, parsed_start, start_is_dir)
     else:
-        start_dir = _pathlib.dirname(parsed_start.path)
-    relpath = _pathlib.relpath(parsed_source.path, start_dir)
-    if not is_path:
-        relpath = relpath.replace("\\", "/")
-    if not relpath.startswith("."):
-        relpath = _join(is_path, ".", relpath)
+        return _make_relative_href_url(parsed_source, parsed_start, start_is_dir)
 
-    return relpath
+
+def _make_absolute_href_url(
+    parsed_source: URLParseResult,
+    parsed_start: URLParseResult,
+    start_is_dir: bool = False,
+) -> str:
+
+    # If the source is already absolute, just return it
+    if parsed_source.scheme != "":
+        return urlunparse((parsed_source))
+
+    # If the start path is not a directory, get the parent directory
+    start_dir = (
+        parsed_start.path if start_is_dir else parsed_start.path.rsplit("/", 1)[0]
+    )
+    # Ensure the directory has a trailing slash so urljoin works properly
+    start_dir = start_dir.rstrip("/") + "/"
```

Comment:
Are you sure you want to do this? It works with \*nix filenames, but in HTTP URLs `foo//bar` and `foo/bar` are semantically distinct. For example, some websites use path separators as some kind of argument separator, as in `https://example.org/api/endpoint/arg1/arg2/…`. It's a terrible practice, but it does happen.

If you want to keep this as is, you might want to `rstrip("/\\")` since this happens before `replace("\\", "/")`. Otherwise `\foo\bar\\` would (I think) end up as `/foo/bar//`, exactly what this is trying to prevent.

Reply:
We need to do this in the case where we've used `.rsplit("/", 1)[0]` (or `dirname` if that ends up working) for `urljoin` to treat it as a directory, but you're right that we shouldn't do it if we are just using `parsed.start.path`.

Reply:
So, it looks like `urljoin` doesn't handle the case that you brought up with empty path arguments in the way we expect:

```python
>>> urljoin("/a/b/x//", "thing")
'/a/b/x/thing'
>>> urljoin("/a/b/x///", "thing")
'/a/b/x/thing'
>>> urljoin("https://a/b/x///", "thing")
'https://a/b/x/thing'
```

Given that this is a pretty standard way of joining URLs in Python I'm inclined to not to try to get around it, but I think I will change the code to the following so that we're at least confident that we are preserving the path properly:

```python
# If the start path is not a directory, get the parent directory
if start_is_dir:
    start_dir = parsed_start.path
else:
    # Ensure the directory has a trailing slash so urljoin works properly
    start_dir = parsed_start.path.rsplit("/", 1)[0] + "/"
```